src/rv32_template.c

```diff
@@ -294,45 +342,58 @@ RVOP(
-    IIF(RV32_HAS(EXT_C))(, const uint32_t pc = PC;);               \
-    if (BRANCH_COND(type, rv->X[ir->rs1], rv->X[ir->rs2], cond)) { \
-        is_branch_taken = false;                                   \
-        struct rv_insn *untaken = ir->branch_untaken;              \
-        if (!untaken)                                              \
-            goto nextop;                                           \
-        IIF(RV32_HAS(JIT))                                         \
-        (                                                          \
-            {                                                      \
-                cache_get(rv->block_cache, PC + 4, true);          \
-                if (!set_add(&pc_set, PC + 4))                     \
-                    has_loops = true;                              \
-                if (cache_hot(rv->block_cache, PC + 4))            \
-                    goto nextop;                                   \
-            }, );                                                  \
-        PC += 4;                                                   \
-        last_pc = PC;                                              \
-        MUST_TAIL return untaken->impl(rv, untaken, cycle, PC);    \
-    }                                                              \
-    is_branch_taken = true;                                        \
-    PC += ir->imm;                                                 \
-    /* check instruction misaligned */                             \
-    IIF(RV32_HAS(EXT_C))                                           \
-    (, RV_EXC_MISALIGN_HANDLER(pc, INSN, false, 0););              \
-    struct rv_insn *taken = ir->branch_taken;                      \
-    if (taken) {                                                   \
-        IIF(RV32_HAS(JIT))                                         \
-        (                                                          \
-            {                                                      \
-                cache_get(rv->block_cache, PC, true);              \
-                if (!set_add(&pc_set, PC))                         \
-                    has_loops = true;                              \
-                if (cache_hot(rv->block_cache, PC))                \
-                    goto end_op;                                   \
-            }, );                                                  \
-        last_pc = PC;                                              \
-        MUST_TAIL return taken->impl(rv, taken, cycle, PC);        \
-    }                                                              \
+#define BRANCH_FUNC(type, cond)                                             \
+    IIF(RV32_HAS(EXT_C))(, const uint32_t pc = PC;);                        \
+    if (BRANCH_COND(type, rv->X[ir->rs1], rv->X[ir->rs2], cond)) {          \
+        is_branch_taken = false;                                            \
```

Comment:
The `is_branch_taken` flag should not be updated when trapped.

Reply:
Yes. Give me some time to test this out.